src/tokenizer/tokenizer.py

```diff
@@ -945,7 +545,46 @@ def parse_tokens(txt):
                 ate = True
                 yield TOK.Number(w[0], unicodedata.numeric(w[0]))
                 w = w[1:]
-
+            if w and w.startswith(URL_PREFIXES):
+                # Handle URL: cut RIGHT_PUNCTUATION characters off its end,
+                # even though many of them are actually allowed according to
+                # the IETF RFC
+                endp = ""
+                while w and w[-1] in RIGHT_PUNCTUATION:
+                    endp = w[-1] + endp
+                    w = w[:-1]
+                yield TOK.Url(w)
+                ate = True
+                w = endp
+            if w and len(w) >= 2 and re.search(r"^#\w", w, re.UNICODE):
+                # Handle hashtags. Eat all text up to next punctuation character
+                # so we can handle strings like "#MeToo-hreyfingin" as two words
+                tag = w[:1]
+                w = w[1:]
+                while w and w[0] not in PUNCTUATION:
+                    tag += w[0]
+                    w = w[1:]
+                if re.search(r"^#\d+$", tag):
+                    # Hash is being used as a number sign, e.g. "#12"
+                    yield TOK.Ordinal(tag, int(tag[1:]))
+                else:
+                    yield TOK.Hashtag(tag)
+                ate = True
+            # Domain name (e.g. greynir.is)
+            if (
+                w
+                and len(w) >= MIN_DOMAIN_LENGTH
+                and w[0].isalnum()  # All domains start with an alphanumeric char
+                and "." in w[1:-2]  # Optimization, TLD is at least 2 chars
+                and (w.startswith("www.") or DOMAIN_REGEX.search(w))
+            ):
+                endp = ""
```

Comment:
Would it not be safer to avoid the `www.` special case, and use only `DOMAIN_REGEX`, and then cut the matched expression from the front of `w` - rather than chopping punctuation off the end of `w`? How would this handle something like `www....`?

Reply:
Yeah, you're right, pointless to check for www. The domain regex should capture everything anyway.